### PR TITLE
add DensifyHaversine trait

### PR DESCRIPTION
**Pull request by JosiahParry**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

This PR adds a `DensifyHaversine` trait which enables users to densify line based spherical geometries. 

It provides the ground work for enabling linestring segmentation on geographical coordinates which can be used by downstream libraries such as [stplanr](https://github.com/ropensci/stplanr/issues/522).

### CODE REVIEW

**Comment by michaelkirk** [comment]
re: the CI failure - You might want to check out our usage of `approx` elsewhere for the sloppy floating point matching.

**Comment by JosiahParry** [comment]
Oh thank goodness. That was an adventure!

**Review by urschrei — APPROVED** [review]
LGTM. I'll leave it open in case anyone else wants to kick the tires.

File: geo/src/algorithm/densify_haversine.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    LineString<T>: HaversineLength<T>,
+{
+    type Output = LineString<T>;
+
+    fn densify_haversine(&self, max_distance: T) -> Self::Output {
+        let mut new_line = vec![];
+        self.lines()
+            .for_each(|line| densify_line(line, &mut new_line, max_distance));
+        // we're done, push the last coordinate on to finish
+        new_line.push(self.points().last().unwrap());
```
Comment (michaelkirk) [inline geo/src/algorithm/densify_haversine.rs:126]: We'll panic here if we have an empty linestring, right? I don't think we want that. Was this copied from the existing densify trait? Seems like we need to fix that too, but we could do it in a separate PR.

I think it's reasonable to return an empty line string when trying to densify an empty line string.

It looks to me like empty geometries of other types are already being handled reasonably.

File: geo/src/algorithm/densify_haversine.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        let dense = linestring.densify_haversine(110.0);
+        assert_relative_eq!(dense, output, epsilon = f64::EPSILON);
+    }
+
+    #[test]
+    fn test_line_densify() {
+        let output: LineString = vec![[0.0, 0.0], [0.0, 0.5], [0.0, 1.0]].into();
+        let line = Line::new(coord! {x: 0.0, y: 0.0}, coord! { x: 0.0, y: 1.0 });
+        let dense = line.densify_haversine(100000.0);
+        assert_relative_eq!(dense, output, epsilon = f64::EPSILON);
```
Comment (michaelkirk) [inline geo/src/algorithm/densify_haversine.rs:249]: Take it or leave it feedback: f64::EPSILON is already the default epsilon. You only need to specify epsilon if you want a default epsilon.

https://docs.rs/approx/latest/src/approx/abs_diff_eq.rs.html#87

**Review by michaelkirk — CHANGES_REQUESTED** [review]

File: geo/src/algorithm/densify_haversine.rs (hunk 1 above)
Comment (JosiahParry) [reply geo/src/algorithm/densify_haversine.rs:126]: This was taken directly from the existing Densify trait. What would you like to do to proceed?

File: geo/src/algorithm/densify_haversine.rs (hunk 2 above)
Comment (JosiahParry) [reply geo/src/algorithm/densify_haversine.rs:249]: I'll take it :)

File: geo/src/algorithm/densify_haversine.rs (hunk 1 above)
Comment (urschrei) [reply geo/src/algorithm/densify_haversine.rs:126]: Something like?

```
fn densify_haversine(&self, max_distance: T) -> Self::Output {
    if self.coords_count() == 0 {
        return LineString::new(vec![]);
    }
    // keep going
```

I can add the same to `densify` as a separate PR

Comment (JosiahParry) [reply geo/src/algorithm/densify_haversine.rs:126]:
```rust

use geo::{LineString, Densify};

fn main() {
  let linestring: LineString<f64> = LineString::new(vec![]);
  linestring.densify(30.0);
}
```
panics as you expected

File: geo/src/algorithm/densify_haversine.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    T: CoordFloat + FromPrimitive,
+    Line<T>: HaversineLength<T>,
+    LineString<T>: HaversineLength<T>,
+{
+    type Output = LineString<T>;
+
+    fn densify_haversine(&self, max_distance: T) -> Self::Output {
+        if self.coords_count() == 0 {
+            return LineString::new(vec![]);
+        }
```
Comment (JosiahParry) [inline geo/src/algorithm/densify_haversine.rs:125]: @michaelkirk I've only added this check here as it seems to be called everywherer else. Should this suffice?

File: geo/src/algorithm/densify_haversine.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        let dense = line.densify_haversine(100000.0);
+        assert_relative_eq!(dense, output);
+    }
+
+    #[test]
+    fn test_empty_linestring() {
+        let linestring: LineString<f64> = LineString::new(vec![]);
+        let dense = linestring.densify_haversine(10.0);
+        assert_eq!(0, dense.coords_count());
+    }
```
Comment (JosiahParry) [inline geo/src/algorithm/densify_haversine.rs:262]: Added an additional test for linestrings. I don't think its possible to have an empty `Line` so i didnt add a test for it.

**Review by michaelkirk — APPROVED** [review]
Thanks!
